frontend/src/app/main/data/workspace/libraries.cljs

```diff
@@ -577,46 +578,84 @@
     ptk/WatchEvent
     (watch [_ state stream]
       (log/info :msg "UPDATE-COMPONENT of shape" :id (str id))
-      (let [local-library (dwlh/get-local-library state)
+      (let [page-id       (get state :current-page-id)
+            local-library (dwlh/get-local-file state)
             libraries     (dwlh/get-libraries state)
+
             [rchanges uchanges]
-            (dwlh/generate-sync-shape-inverse (get state :current-page-id)
+            (dwlh/generate-sync-shape-inverse page-id
                                               id
                                               local-library
-                                              libraries)]
-
-        (log/debug :msg "UPDATE-COMPONENT finished" :js/rchanges (log-changes
-                                                                   rchanges
-                                                                   local-library))
-
-        (rx/of (dwc/commit-changes rchanges uchanges {:commit-local? true}))))))
+                                              libraries)
+
+            container (cp/get-container page-id :page local-library)
+            shape     (cp/get-shape container id)
+            file-id   (:component-file shape)
+            file      (dwlh/get-file state file-id)
+
+            local-rchanges (->> rchanges
+                               (filter :local-change?)
+                               (map #(dissoc % :local-change?))
+                               vec)
+            local-uchanges (->> uchanges
+                               (filter :local-change?)
+                               (map #(dissoc % :local-change?))
+                               vec)
```

Comment:
Here we have a duplicated seq processing logic that can be expressed better using reusable transducers, that improves reusability and performance:

```clojure
      xf-filter      (comp
                      (filter :local-change?)
                      (map #(dissoc % :local-change?)))
      local-rchanges (into [] xf-filter rchanges)
      local-uchanges (into [] xf-filter uchanges)
```

Reply:
done